broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java

```diff
@@ -0,0 +1,222 @@
+
+  @Rule
+  public final RuleChain chain =
+      RuleChain.outerRule(temporaryFolder).around(actorScheduler).around(logStorageRule);
+
+  private AtomixSnapshotStorage snapshotStorage;
+  private AtomixLogStorageReader storageReader;
+  private LogDeletionService deletionService;
+  private Compactor compactor;
+
+  @Before
+  public void setUp() {
+    storageReader =
+        new AtomixLogStorageReader(logStorageRule.getRaftLog().openReader(-1, Mode.COMMITS));
+    snapshotStorage =
+        new AtomixSnapshotStorage(
+            null,
+            logStorageRule.getSnapshotStore(),
+            new AtomixRecordEntrySupplierImpl(storageReader),
+            1,
+            new SnapshotMetrics(PARTITION_ID));
+    compactor = new Compactor();
+
+    deletionService = new LogDeletionService(0, PARTITION_ID, compactor, snapshotStorage);
+    actorScheduler.submitActor(deletionService).join();
+  }
+
+  @After
+  public void tearDown() {
+    if (storageReader != null) {
+      storageReader.close();
+      storageReader = null;
+    }
+
+    if (snapshotStorage != null) {
+      snapshotStorage.close();
+      snapshotStorage = null;
+    }
+
+    if (deletionService != null) {
+      deletionService.close();
+      deletionService = null;
+    }
+  }
+
+  @Test
+  public void shouldDeleteUpToCompactionBound() {
+    // given
+    final var reader = logStorageRule.getRaftLog().openReader(-1);
+
+    // when
+    logStorageRule.appendEntry(1, 1, DATA).index();
+    logStorageRule.appendEntry(2, 2, DATA).index();
+    logStorageRule.appendEntry(3, 3, DATA).index();
+    createSnapshot(2);
+
+    // then
+    compactor.awaitCompaction(2L, Duration.ofSeconds(5));
+    reader.reset();
+    final var entries = readAllEntries(reader);
+    assertThat(entries).isNotEmpty().hasSize(2).extracting(Indexed::index).containsExactly(2L, 3L);
+  }
+
+  @Test
+  public void shouldNotDeleteOnLowerCompactionBound() {
+    // given
+    final var reader = logStorageRule.getRaftLog().openReader(-1);
+
+    // when
+    logStorageRule.appendEntry(1, 1, DATA).index();
+    logStorageRule.appendEntry(2, 2, DATA).index();
+    logStorageRule.appendEntry(3, 3, DATA).index();
+    createSnapshot(0);
+
+    // then
+    compactor.awaitCompaction(0L, Duration.ofSeconds(5));
+    reader.reset();
+    final var entries = readAllEntries(reader);
+    assertThat(entries)
+        .isNotEmpty()
+        .hasSize(3)
+        .extracting(Indexed::index)
+        .containsExactly(1L, 2L, 3L);
+  }
+
+  @Test
+  public void shouldDeleteLowerEntriesEvenIfIndexNotFound() {
+    // given
+    final var reader = logStorageRule.getRaftLog().openReader(-1);
+
+    // when
+    logStorageRule.appendEntry(1, 1, DATA).index();
+    logStorageRule.appendEntry(2, 2, DATA).index();
+    logStorageRule.appendEntry(3, 3, DATA).index();
+    createSnapshot(5L);
+
+    // then - expect exactly one segment left with entry 3
+    compactor.awaitCompaction(5L, Duration.ofSeconds(5));
```

Comment:
Not sure whether this makes sense or not. This would mean if we get a snapshot with index 5 we keep 3, so actually we have an gap in between. Hopefully we will start not  at the beginning :sweat_smile:

Reply:
This is already the case, though - I just replicated current feature parity :sweat_smile: This is more a side effect of the journal design in Atomix how we use it - there is always one segment left, so even if you try to compact because you received a snapshot ahead of your log, you'll always have some data left over =/

Reply:
but with truncate we delete it right?

Reply:
I'll double check, but afaik in Atomix, on receiving a snapshot, we reset the head to that index (meaning we truncate down to that index), and start appending. So if we had 1, 2, 3, and we receive snapshot 25, then we truncate nothing and the log will be 1,2,3,26,27, etc.

Reply:
Hm so we can't start from a clean state